src/bp/core/services/workspace-service.ts

```diff
@@ -162,18 +162,28 @@ export class WorkspaceService {
     return this.workspaceRepo.getWorkspaceUsers(workspace)
   }
 
-  async getWorkspaceUsersAttributes(workspace: string, filteredAttributes?: string[]): Promise<StrategyUser[]> {
-    const users = await this.workspaceRepo.getWorkspaceUsers(workspace)
+  async getWorkspaceUsersAttributes(
+    workspace: string,
+    filteredAttributes?: string[]
+  ): Promise<WorkspaceUserAttributes[]> {
+    const workspaceUsers = await this.workspaceRepo.getWorkspaceUsers(workspace)
 
-    return _.flatten(
-      await Promise.map(_.uniq(_.map(users, 'strategy')), strategy => {
-        return this.usersRepo.getMultipleUserAttributes(
-          users.filter(x => x.strategy === strategy).map(x => x.email),
-          strategy,
-          filteredAttributes
+    const uniqStrategies = _.uniq(_.map(workspaceUsers, 'strategy'))
+
+    const usersInfo = _.keyBy(
+      _.flatten(
+        await Promise.map(uniqStrategies, strategy =>
+          this.usersRepo.getMultipleUserAttributes(
+            workspaceUsers.filter(x => x.strategy === strategy).map(x => x.email),
+            strategy,
+            filteredAttributes
+          )
         )
-      })
+      ),
+      'email'
```

Comment:
I know you didn't write this block, but I find there's a lot going on here.

Reply:
Yeah you're right. It needs a refactor.